packages/replay/src/auth.ts

```diff
@@ -0,0 +1,265 @@
+}
+
+async function refresh(refreshToken: string) {
+  try {
+    const resp = await fetch(`https://${getAuthHost()}/oauth/token`, {
+      method: "POST",
+      headers: { "Content-Type": "application/json" },
+      body: JSON.stringify({
+        audience: "https://api.replay.io",
+        scope: "openid profile",
+        grant_type: "refresh_token",
+        client_id: getAuthClientId(),
+        refresh_token: refreshToken,
+      }),
+    });
+
+    const json = await resp.json();
+
+    if (json.error) {
+      debug("OAuth token request failed: %O", json.error);
+
+      throw {
+        id: "auth0-error",
+        message: json.error,
+      };
+    }
+
+    if (!json.access_token) {
+      debug("OAuth token request was missing access token: %O", json);
+
+      throw {
+        id: "no-access-token",
+      };
+    }
+
+    return json.access_token;
+  } catch (e: any) {
+    throw {
+      ...e,
+      refreshToken: maskToken(refreshToken),
+    };
+  }
+}
+
+function generateAuthKey() {
+  const hash = createHash("sha256");
+  hash.write(String(globalThis.performance.now()));
+  return hash.digest("hex").toString();
+}
+
+function initAuthRequest(options: Options = {}) {
+  maybeLog(options.verbose, "🌎 Launching browser to login to replay.io");
+  const key = generateAuthKey();
+  const server = process.env.REPLAY_APP_SERVER || "https://app.replay.io";
+  spawn(openExecutable(), [`${server}/api/browser/auth?key=${key}&source=cli`]);
+
+  return key;
+}
+
+async function fetchToken(key: string) {
+  const resp = await query(
+    "CloseAuthRequest",
+    `
+        mutation CloseAuthRequest($key: String!) {
+          closeAuthRequest(input: {key: $key}) {
+            success
+            token
+          }
+        }
+      `,
+    {
+      key,
+    }
+  );
+
+  if (resp.errors) {
+    if (
+      resp.errors.length === 1 &&
+      resp.errors[0].message === "Authentication request does not exist"
+    ) {
+      throw {
+        id: "missing-request",
+      };
+    } else {
+      throw {
+        id: "close-graphql-error",
+        message: resp.errors
+          .map((e: any) => e.message)
+          .filter(Boolean)
+          .join(", "),
+      };
+    }
+  } else if (!resp.data.closeAuthRequest.token) {
+    // there's no obvious reason this would occur but for completeness ...
+    throw {
+      id: "close-missing-token",
+      message: JSON.stringify(resp),
+    };
+  }
+
+  const refreshToken = resp.data.closeAuthRequest.token;
+
+  return refreshToken;
+}
+
+export async function pollForToken(key: string, options: Options = {}) {
+  let timedOut = false;
+  setTimeout(() => {
+    timedOut = true;
+  }, 60 * 1000);
+
+  while (true) {
+    if (timedOut) {
+      debug("Timed out waiting for auth request");
+      throw { id: "timeout" };
+    }
+
+    try {
+      const refreshToken = await fetchToken(key);
+      maybeLog(options.verbose, "🔑 Fetching token");
+
+      return await refresh(refreshToken);
```

Comment:
Why do we use a refresh token here? Why doesn't `CloseAuthRequest` simply return the access token?

Reply:
We only store the refresh token in the db right now and exchange it for a token right away so the refresh token is invalidated and we're not holding into anything that could be used to impersonate the user.